### ISSUE TITLE
4.23.0 introduced a breaking change in `Paths` type

### ISSUE THREAD

**Issue by mwaibel-go** [sindresorhus/type-fest#922]
I get this error with type-fest 4.23.0 that I didn’t get with 4.22.1:

```
Type 'string' is not assignable to type 'never'.(2322)
main.ts(3, 3): The expected type comes from property 'column' which is declared here on type 'ColumnSortState<Record<any, any>>'
```

Offending code ([Stackblitz](https://stackblitz.com/edit/vitejs-vite-m37qbs?file=index.html,src%2Fmain.ts,tsconfig.json&terminal=dev)):
```ts
import { Paths } from "type-fest";
interface ColumnSortState<Sortee extends object> {
  column: Paths<Sortee>;
}


function sort<Sortee extends Properties, Properties extends Record<any, any>>(
  _sortees: Sortee[],
  _sortState: ColumnSortState<Properties>,
): void {
}


interface Dto {
  startDate: string
}
let dtos!: Dto[]
sort(
  dtos,
  {
    column: 'startDate',
  }
)

```

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/922">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/922/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/sindresorhus/type-fest/issues/922/pledge.svg">
</picture>
</a>

**Comment by sindresorhus** [comment]
// @fardolieri

**Comment by voxpelli** [comment]
The change causing the regression: https://github.com/sindresorhus/type-fest/pull/920

**Comment by voxpelli** [comment]
Why do `sort<Sortee extends Properties, Properties extends Record<any, any>>` rather than just `sort<Sortee extends Record<any, any>>`?

**Comment by fardolieri** [comment]
Oh that is really interesting. I created a trimmed down version that shows the problem [Playground Link](https://www.typescriptlang.org/play/?ts=5.5.3#code/FAMwrgdgxgLglgewgAgM4IE4wPIBsAmAPAILICmAHjGRPqsgEIA0jAfABQCGAXMsSwCNeABU4wAFqjxEGrAJS8Abgjj5kAbwC+wdFmnt1yEAgS8A5GeSaWZ4wjNydmHAQNGT5y9eRmBnDA5Oeq6GdrzqfhiemlY2dgB0kYGgkLCIKLowAHJkAO4k5FQ0dIwssly8-MhCyKISqDn5sgrIyqoa2pmNbmE+XnEmgV15PR59sT5JjsO5o6YakdETtiaJ-snAAPSbyAAq4mTIUJgYZLDIuZwAnsgwCMhgqId1ksAp0PBIaM4AwidnLiIxA4PD4ghEYkk0hI8iUKjUWiCMD+GFOsH0oTGFmWdiGv3+6JC7nm2O8vnW03xqIBGOJ4UWfRiZISUze4A+6W+WBRaOyeRhFTB1Qh9UaMJabQRnSpvO6mJJ-R8uMp3IJfNm8s8y1ZmR5ALldIW-iWzNWrLebzgAFsAA7OW5XG2HdQAZRgYjgUFEehAxFR1xYADV-HBOAJcGRvThff6riwskgAEpnMAYVBwRRkXaOsgsXYIN0YOAQADmMRAGAQVp8MBzAFoQGRUDBNuhU1AyJti9QMBBOLgzABuYDWu1YB1OjQAUVttewAgAVgDy5Xq2Za06G02YEOR7b7RvnQBJVDECBXFdVmv1xvN3ejg85jQnnKZjCXteHrd34cP8eHjQAFUIAAawgBBcggP0MGuD9r03W8d1-fd-yfV0wAEGAYNgOD1xvbd7xQmAJ2dABxU4xDIDB9k4CBcK-RDCLHYiAPUAAZJtUBouijFXeCyG-JC92YkiNBdMArXo-CfzeSgRIAl4pBtT4IHoABeDRgAASCtTgKGTKBU3TJAABEyGU8QAH5eAgCSBCo4dNGHYA5MfSdFLFXYWGwZT0noShqFoehFJ8lT1I6VhkDU7TdkKQKSgTCADKMjMsyfAAfZBk04fAkFwK4AFlOBtQhIDAiCIBYMrwMgyLMuy3KIHyl0yBgUrQJqiBWG0rTLOQCAyDfHreBPM8rkIXZIoC4p6CwsAyB63r+sGqjFt4WLpqC5BgPKyDoOuRaltsq17IwOKZr2ABtMwI1LCQzAAXUOrTtmQAB1Q4BrINQ7mQcRaPwCNbgOIw4Aob7kAB-qkAbMGIeLfBKGQG1-GIhAQGBw5-Bgm4nhRmDqHy+Jnr6o8IB7PtcA8-k3Q9L1UewGMcYm1hvN8pBUG6rTue5zKyYp-tqfyfbxuDIswwjKNGZFlnLuO06HtZ5BQr8rmea0kbyaoymhYmtmwrV7n1vOraEEXAESeQfntcFyEGn5LzlfZ1TDaN5a32cgDrd7W3RQd-W-JNkoQud8KtEi6KtPYFWOeu3T9JTNN0jMizHqD+h5ao5A+pj1S4705Kk9M8z7oe5BeAARgABjkdPkGLRszqKigU4kOvM4wHq+uTABHMA4FOIhJrrhus92Ra+o2ootpnZSrnnJdzkyijGvy5BLqe9W+oGob1fCZ7LoAaTIG5i2QECT-RvYHu4Z7j5uTaSmbItS2QTKO+QV62OtOBiIAA3vn-W49xn7FhLMgTA-U7JUWJurbmr0jwYwASfIBcB6CcCgSdKiLBTgwFTCgZBNxMp-wACTqHvpoP+f9BBnE4I8Q4AgEBtz-tja4l0q4PSAVDFhsZrpV0ekA3ImAQKwLgZZZ6WlMr3wkZlfMhYwGEHvq7XmyB2ASK0uRMglFqL-QgIQZurdxAsCrlNaeJQ5qHC-j-FiINEYWWAcjU4mZyb1wgCAYsv9DhohSkgdRS0XwrQwIQXWuwj4nweiwdQ8dC7GQgIY3g4lMLYTagYkuRjkAV1YJoVgpj4r0BAP2J4filqkPISfTQ8QyEhLCVcCJGhomJ1ifE5AiSsKcFgPovShiWCZOyZQ4pNlAl+MGbvOBjg4Ea3dqtdWmhLoXyuFfWKAAyVRU88nbQ6hVEW2dMGnXLg8TZkE5CbzdjvaZkyzkYGcq5VC7k7bQmHhpEKBAAD6LNPZPmefgN5jtDHt2gWdDSJiooxTrolGJqVsyTnqlo1ehVirtV2pVA5SK6pZVhXlK4LU2rVQqmrbeQy3ajXPCzOuFiJ5TM7jzY2j96A7U6iLQ628AV11CTdGgJZS7PVeh9ZaENfq6MBocCQhx3HgzUFDcCEBYbitcYjCgyNUYQIxiKyGsY0DmX8JRImltvY63uQQQgtN4D0x9DLSaLBDHKL5lrH2VMDVAljIQMWoZwyRgZkza4ssO6K0tWk5RmsBb2vqA8v1FlXY0rMfQM2i8YC6ttfqkNhrfn+ueiMhyskKDySfHq32UJDUvJTfY2luys5ArzFFdFfcB7fUIIWnJIKtJrIurPOc5sl7IBXpi9eJyCWjP3jzGprjz6XwxrsG+i17511Aa-d+LKrFWl-sgQhQDfozvAZAjuoj4E7EQcuwB9d0GlowDg1q+D90nzfsushFCqE0KgHQp41QmHiGXawq47DOGQ1oG+3hZh+FfqERgERTLDpSJPmBvYBYsIKKUZBtRcCOKoC4rowg3SMkAFZckXQsZ-HY39F02MOHYtuv0bROJoMRBuHjqCdsaekCRpN7ZvmCQ6n5NS6niStGhtJPScnYa2gU3ARSJklJveUyp6gvnsfvpxiSPGLJ8ayX-dR6aqVwLU4dcZ1LKU9VmfMxZyAVnsGbVtelWz1XMqwWdXguKjkPUHEAA)

The example of @mwaibel-go only worked with the old version of `Paths` because they abused a side effect where a (strange looking) Type could be inferred from a given input path. 

If we look at the example `sortOld({ foo: '' }, 'bar')` then the error message is
> 'foo' does not exist in type '{ bar: any; } & { bar: any; } & { bar: any; } & { bar: any; } & { bar: any; } & { bar: any; } & { bar: any; } & { bar: any; }'.

Meaning that typescript could go the reverse direction and create a type `{bar: any}` (simplified) from the input `'bar'`. But this breaks on the first attempt of nesting, see `sortOld({ foo: {bar: ''} }, 'foo.bar')`. The type being inferred from `foo.bar` is `{ "foo.bar": any; } & { foo: any; }` which is missing `{ foo: { bar: any } }`. That means the example never even worked with nested paths, which is the whole point of `Paths`. If you only need "one layer" you can use `keyof T`.

**TLDR;** The way @mwaibel-go used `Paths<T>` only allowed keys that are one layer deep, which is the same as `keyof T`. If you want the exact same behavior as before:

```diff
import { Paths } from "type-fest";
interface ColumnSortState<Sortee extends object> {
- column: Paths<Sortee>;
+ column: keyof Sortee;
}


function sort<Sortee extends Properties, Properties extends Record<any, any>>(
  _sortees: Sortee[],
  _sortState: ColumnSortState<Properties>,
): void {}
```

If you actually want to use `Paths<T>` and allow multi layered keys:

```ts
function sort<Sortee extends Record<any, any>>(
  _sortees: Sortee[],
  _sortState: ColumnSortState<Sortee>,
): void {}
```

**Comment by bleistift-zwei** [comment]
> Why do `sort<Sortee extends Properties, Properties extends Record<any, any>>` rather than just `sort<Sortee extends Record<any, any>>`?

It’s a minimal example. The original function requires access to both `Sortee` and `Properties`.

**Comment by mwaibel-go** [comment]
I’m sorry for the noise. I cannot reproduce the error. Thanks for pointing out the issue in my code, @fardolieri!
